ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -1295,4 +998,533 @@ private void endCompaction() {
+      // anyway.
+      //
+      // This is a table that shows how the validity of the current IndexValue is determined.
+      // ----------------------------------------------------------------------------------------
+      // Current IndexValue  | Latest IndexValue | Action                                       |
+      // --------------------+-------------------+----------------------------------------------|
+      // Put(verion c)       | Put(version f)    | isExpired(Pc)?false:true                     |
+      //                     | Delete(f)         | reachRetention(Df)||isExpired(Df)?false:true |
+      //                     | Undelete(f)       | isExpired(Uf)?false:true                     |
+      // --------------------+-------------------+----------------------------------------------|
+      // TtlUpdate(c)        | Put(f)            | Exception                                    |
+      //                     | Delete(f)         | reachRetention(Df)?false:true                |
+      //                     | Undelete(f)       | true                                         |
+      // --------------------+-------------------+----------------------------------------------|
+      // Delete(c)           | Put(f)            | Exception                                    |
+      //                     | Delete(f)         | c==f?true:false                              |
+      //                     | Undelete(f)       | false                                        |
+      // --------------------+-------------------+----------------------------------------------|
+      // Undelete(c)         | Put(f)            | Exception                                    |
+      //                     | Delete(f)         | false                                        |
+      //                     | Undelete(f)       | c==f&&!isExpired(Uc)?true:false              |
+      // ----------------------------------------------------------------------------------------
+      List<IndexEntry> validEntries = new ArrayList<>();
+      Iterator<IndexEntry> iterator = indexSegment.getIterator();
+      StoreKey lastKey = null;
+      IndexValue lastFinalState = null;
+      while (iterator.hasNext()) {
+        IndexEntry entry = iterator.next();
+        StoreKey currentKey = entry.getKey();
+        IndexValue currentValue = entry.getValue();
+        IndexValue currentFinalState;
+
+        if (lastKey == null) {
+          // save a token for restart (the key gets ignored but is required to be non null for construction)
+          StoreFindToken safeToken =
+              new StoreFindToken(currentKey, indexSegment.getStartOffset(), sessionId, incarnationId);
+          compactionLog.setSafeToken(safeToken);
+          logger.debug("Set safe token for compaction in {} to {}", storeId, safeToken);
+        }
+        if (currentKey.equals(lastKey)) {
```

Comment:
When will this happen?

Reply:
This method would iterate through all the IndexValue in an IndexSegment. For same key, they might be multiple IndexValue, like. PUT + TTL + DELETE, since IndexSegment sorts IndexValues based on their keys, so if we have multiple IndexValues for the same key in the same IndexSegment, then they must follow each other. So here, if we detect the current key equals to the previous one, we don't have to search for the final state for the same key again.

Reply:
Makes sense. Better to add a comment here.

Reply:
+1 , comment to explain why don't need to query latest state: because they share same latest state.